[FOCAL-66] ECAL geoemetry and step manager for pads

amatyja opened the pull request:
[FOCAL-66] ECAL geometry and step manager for pads

amatyja committed:
[FOCAL-66] ECAL geometry and step manager for pads

[FOCAL-66] ECAL geometry and step manager for pads

amatyja committed:
[FOCAL-66] ECAL geometry and step manager for pads

amatyja committed:
Update Detector.cxx with whitespaces

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -165,6 +184,54 @@ Bool_t Detector::ProcessHits(FairVolume* v)
+    LOG(debug) << "We are in sensitive volume " << v->GetName() << ": " << TVirtualMC::GetMC()->CurrentVolPath();
+
+    double eloss = TVirtualMC::GetMC()->Edep() * 1e9; // energy in eV  (GeV->eV)
+    if (eloss < DBL_EPSILON) {
+      return false; // only process hits which actually deposit some energy in the FOCAL
+    }
```
Could be potentially dangerous, as it might shadow other parts (i.e. pixel simulation which comes later). I think we need to outsource the functions (also for HCAL @hahassan7) into dedicated process functions.

amatyja replied:
ProcessHit method is split into two parts. It contains two submethods for ECAL and HCAL, separatelly.

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -165,6 +184,54 @@ Bool_t Detector::ProcessHits(FairVolume* v)
   }
 
+  // Processing ECAL hits
+  if (TVirtualMC::GetMC()->CurrentMedium() == mMedSensECal) {
```
Please make sure that pad and pixel layers are declared as separate sensitive volumes. Because the pixels need a different hit simulation (mixture of Detector simulation in ITSMFT and the gaussian diffusion model)

amatyja replied:
Pixels and Pads are separated in geometry and in process hit method.

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -165,6 +184,54 @@ Bool_t Detector::ProcessHits(FairVolume* v)
+      return true;
+    }
+
+    auto currenthit = FindHit(mCurrentParentID, col, row, layer);
```
Please cross check that HCAL and ECAL layers are treated as separate layers in FindHit

amatyja replied:
I belive that the method getVirtualInfo(...) from Geometry class is doing the job properly.

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -418,9 +485,31 @@ void Detector::addAlignableVolumesHCAL() const
+//____________________________________________________________________________
+void Detector::addAlignableVolumesECAL() const
+{
+  TString vpsector = "/cave_1/caveRB24_1/FOCAL_1/ECAL_1";
+  TString snsector = "FOCAL/ECAL";
```
Please avoid using classes from ROOT unless needed. TString in this case should be converted to std::string

amatyja replied:
Corrected with std:string in both HCAL and ECAL

mfasDa reviewed (changes requested):
I think the most important change needed is to outsource the Hit processing of the separate subsystems to dedicated functions ProcessHitsEPad, ProcessHitsEPixel, ProcessHitsHCAL, for 2 reasons:
a) The check for minimal energy deposit might shadow other subsystems, if they are in a combined process function
b) We would like to be able to switch off subsystems via the o2::ConfigurableParams class to run ECAL or HCAL alone (i.e. for testbeam simulation)

Otherwise I think the PR is already in a good shape and soon ready to be merged.

alibuild commented:
Error while checking build/O2/fullCI for 547b62e90dd682eb2d7d3548a63a12afc004fbfc at 2024-07-04 09:30:
```
## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/13249-slc8_x86-64/0/Detectors/FOCAL/simulation/src/Detector.cxx:980:33: error: use nullptr [modernize-use-nullptr]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13249/547b62e90dd682eb2d7d3548a63a12afc004fbfc/build_O2_fullCI/pretty.html).

amatyja committed:
Update Detector.cxx - use nullptr

amatyja committed:
Update Detector.cxx with sdt:string

amatyja committed:
Update Detector.h with split ProcessHit for ECAL and HCAL

Added ProcessHItsECAL and ProcessHitHCAL methods.

amatyja committed:
Update Detector.cxx with split ProcessHit for ECAL and HCAL part

Split ProcessHit to ProcessHitECAL and ProcssHitHCAL.

amatyja committed:
Update Detector.h - split Pad and Pixels geometry and process hits

amatyja committed:
Detector.cxx update with split ECAL Pad and Pixel geometry and ProcessHits 

Added sensitive volume for Pixels. Minor changes in geometry - splitting of Pads and Pixels. Additional method for Pixels responsible for ProcessHits in Pixels.

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -453,7 +471,14 @@ void Detector::ConstructGeometry()
 
   TVirtualMC::GetMC()->Gsvolu("FOCAL", "BOX", getMediumID(ID_AIR), pars, 4);
   mSensitiveHCAL.push_back("FOCAL");
+  mSensitiveECALPad.push_back("FOCAL");
+  mSensitiveECALPix.push_back("FOCAL");
```
I think it is enough to have one container for sensitive volumes. The idea behind it is to declare all volumes as sensitive in the framework (so the FairRoot step manager calls ProcessHits for every step allowing us to build the decay chain). Separate lists would only make sense to distinguish in the second step of ProcessHits the "real" sensitive volumes for calling the proper ProcessHits function of the 3 subsystems.

Like this you declare the top volume 3 times.

amatyja replied:
Hi Markus, Do you suggest to create one container for the purpose of geometry creation and separate containers to keep ECAL Pads or Pixels or HCAL for the step manager?

mfasDa replied:
Indeed this is the idea. The reason is that we need to call ProcessHits for every step to build the tree of decay chains, as we cannot access them by the stack, and if we would do this only for the real sensitive volumes we would get broken chains - we saw this in EMCAL. For this purpose we don't distinguish between the subdetectors and can use a common list. Once this is done we do the separate volume check ourselves and redirect to separate ProcessHits functions for Pads, Pixels and HCAL based on the real sensitive volume ID.

mfasDa reviewed (changes requested):
There is one small issue left:

The vector "mSensitive..." we use to declare all FOCAL volumes as sensitive - in order to build the decay tree and not to have broken chains. For this one single container is sufficient - currently the top volume is declared 3 times as sensitive, for each subsystem. The real sensitive volume, for which hits are created, is anyhow handled differently.

amatyja committed:
Update Detector.h - one sensitive volume list

One sensitive volume list instead of three for HCAL, Pads Pixels.

amatyja committed:
Update Detector.cxx - one sensitive volume list

amatyja committed:
Update Detector.h - clang-format

mfasDa reviewed (dismissed):
Now looks good and ready to merge.

amatyja committed:
Update Detector.cxx - added missing string operations

amatyja commented:
I've added missing string method c_srt() to logs.

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -406,7 +406,7 @@ void Detector::addAlignableVolumesHCAL() const
 
   if (!gGeoManager->SetAlignableEntry(snsector.c_str(), vpsector.c_str())) {
-    LOG(fatal) << (Form("Alignable entry %s not created. Volume path %s not valid", snsector, vpsector));
+    LOG(fatal) << (Form("Alignable entry %s not created. Volume path %s not valid", snsector.c_str(), vpsector.c_str()));
```
Please use the streaming operator directly
```c++
LOG(fatal) << "Alignable entry " << snsector << " not created. Volume path " << vpsector << " not valid";
```

amatyja replied:
Done

mfasDa commented on Detectors/FOCAL/simulation/src/Detector.cxx:
```diff
@@ -417,7 +417,7 @@ void Detector::addAlignableVolumesECAL() const
 
   if (!gGeoManager->SetAlignableEntry(snsector.c_str(), vpsector.c_str())) {
-    LOG(fatal) << (Form("Alignable entry %s not created. Volume path %s not valid", snsector, vpsector));
+    LOG(fatal) << (Form("Alignable entry %s not created. Volume path %s not valid", snsector.c_str(), vpsector.c_str()));
```
Also here please use the standard practice for streams (see above).

amatyja replied:
Done

amatyja committed:
Update Detector.cxx - use directly the stream operator in LOG

mfasDa reviewed (dismissed):
Now OK, ready to merge

amatyja committed:
Update Detector.cxx - fix missing return in function
